### PR TITLE
85 implement video service and features

### PR DESCRIPTION
**Pull request by 4thlabs**

### CODE REVIEW

File: src/DolbyIO.Comms.Sdk/Native/Structs/MediaConstraints.cs [hunk 1]
```diff
@@ -16,6 +16,14 @@ public class MediaConstraints
         [MarshalAs(UnmanagedType.U1)]
         public bool Audio = false;
 
+        /// <summary>
+        /// A boolean that indicates whether the application should
+        /// capture the local video and send it to the conference.
+        /// </summary>
+        /// <returns>If true, the SDK will capture video and send it to the conference.</returns>
```
Comment (FabienLavocat) [inline src/DolbyIO.Comms.Sdk/Native/Structs/MediaConstraints.cs:23]: `the SDK will capture video` => `the SDK will capture the local video`

File: src/DolbyIO.Comms.Sdk/Native/Structs/ParticipantInfo.cs [hunk 2]
```diff
@@ -26,7 +26,7 @@ public class ParticipantInfo
         public readonly string Name = "";
         
         /// <summary>
-        /// The URL of the participant's avatar.
+        /// The URL of the participant's avatar.`
```
Comment (FabienLavocat) [inline src/DolbyIO.Comms.Sdk/Native/Structs/ParticipantInfo.cs:29]: Please remove the last `

File: src/DolbyIO.Comms.Sdk/Native/Structs/VideoDevice.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    /// <summary>
+    /// The Video class contains a platform-agnostic description of a video device.
+    /// </summary>
+    [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Ansi)]
+    public struct VideoDevice
+    {
+        /// <summary>
+        /// The unique identifier of the audio device.
+        /// </summary>
+        /// <returns>The identifier.</returns>
```
Comment (FabienLavocat) [inline src/DolbyIO.Comms.Sdk/Native/Structs/VideoDevice.cs:14]: `The identifier of the device.` (to keep it consistent with the next property)

File: src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs [hunk 4]
```diff
@@ -15,48 +15,93 @@ namespace DolbyIO.Comms.Services
...
+        private AudioDeviceAddedEventHandler _audioAdded;
 
         /// <summary>
-        /// Sets the <see cref="DeviceAddedEventHandler"/> that is raised when a new audio or video device is added to the system.
+        /// Sets the <see cref="AudioDeviceAddedEventHandler"/> that is raised when a new audio device is added to the system.
         /// </summary>
-        /// <value>The <see cref="DeviceAddedEventHandler"/> event handler.</value>
-        public DeviceAddedEventHandler Added
+        /// <value>The <see cref="AudioDeviceAddedEventHandler"/> event handler.</value>
+        public AudioDeviceAddedEventHandler AudioAdded
```
Comment (FabienLavocat) [inline src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs:24]: Could we use `AudioDeviceAdded`, without `Device` it sounds like you are getting an audio stream.

File: src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs [hunk 5]
```diff
@@ -15,48 +15,93 @@ namespace DolbyIO.Comms.Services
...
+        private AudioDeviceRemovedEventHandler _audioRemoved;
 
         /// <summary>
-        /// Sets the <see cref="DeviceRemovedEventHandler"/> that is raised when an audio or video device is removed from the system.
+        /// Sets the <see cref="AudioDeviceRemovedEventHandler"/> that is raised when an audio device is removed from the system.
         /// </summary>
-        /// <value>The <see cref="DeviceRemovedEventHandler"/> event handler.</value>
-        public DeviceRemovedEventHandler Removed
+        /// <value>The <see cref="AudioDeviceRemovedEventHandler"/> event handler.</value>
+        public AudioDeviceRemovedEventHandler AudioRemoved
```
Comment (FabienLavocat) [inline src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs:39]: Could we use `AudioDeviceRemoved`

File: src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs [hunk 6]
```diff
@@ -15,48 +15,93 @@ namespace DolbyIO.Comms.Services
...
+        private AudioDeviceChangedEventHandler _audioChanged;
 
         /// <summary>
-        /// Sets the <see cref="DeviceChangedEventHandler"/> that is raised when the currently used input or output device has changed.
+        /// Sets the <see cref="AudioDeviceChangedEventHandler"/> that is raised when the currently used input or output audio device has changed.
         /// </summary>
-        /// <value>The <see cref="DeviceChangedEventHandler"/> event handler.</value>
-        public DeviceChangedEventHandler Changed
+        /// <value>The <see cref="AudioDeviceChangedEventHandler"/> event handler.</value>
+        public AudioDeviceChangedEventHandler AudioChanged
```
Comment (FabienLavocat) [inline src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs:54]: Could we use `AudioDeviceChanged`

File: src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs [hunk 7]
```diff
@@ -15,48 +15,93 @@ namespace DolbyIO.Comms.Services
...
+            }
+        }
+
+        private VideoDeviceAddedEventHandler _videoAdded;
+
+        /// <summary>
+        /// Sets the <see cref="VideoDeviceAddedEventHandler"/> that is raised when a new video device is added to the system.
+        /// </summary>
+        /// <value>The <see cref="AudioDeviceAddedEventHandler"/> event handler.</value>
+        public VideoDeviceAddedEventHandler VideoAdded
```
Comment (FabienLavocat) [inline src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs:69]: Could we use `VideoDeviceAdded`

File: src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs [hunk 8]
```diff
@@ -15,48 +15,93 @@ namespace DolbyIO.Comms.Services
...
+            }
+        }
+
+        private VideoDeviceChangedEventHandler _videoChanged;
+
+        /// <summary>
+        /// Sets the <see cref="VideoDeviceChangedEventHandler"/> that is raised when the currently used input or output video device has changed.
+        /// </summary>
+        /// <value>The <see cref="VideoDeviceChangedEventHandler"/> event handler.</value>
+        public VideoDeviceChangedEventHandler VideoChanged
```
Comment (FabienLavocat) [inline src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs:84]: Could we use `VideoDeviceChanged`

File: src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs [hunk 9]
```diff
@@ -15,48 +15,93 @@ namespace DolbyIO.Comms.Services
...
+            }
+        }
+
+        private VideoDeviceRemovedEventHandler _videoRemoved;
+
+        /// <summary>
+        /// Sets the <see cref="VideoDeviceRemovedEventHandler"/> that is raised when a video device is removed from the system.
+        /// </summary>
+        /// <value>The <see cref="VideoDeviceRemovedEventHandler"/> event handler.</value>
+        public VideoDeviceRemovedEventHandler VideoRemoved
```
Comment (FabienLavocat) [inline src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs:99]: Could we use `VideoDeviceRemoved`

File: src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs [hunk 10]
```diff
@@ -134,5 +179,45 @@ public async Task SetPreferredAudioOutputDeviceAsync(AudioDevice device)
...
             await Task.Run(() => Native.CheckException(Native.SetPreferredAudioOutputDevice(device))).ConfigureAwait(false);
         }
+
+        /// <summary>
+        /// Gets a list of all video devices that are currently available in the system.
+        /// </summary>
+        /// <returns>The <xref href="System.Threading.Tasks.Task`1"/> that represents the asynchronous operation.
+        /// The <xref href="System.Threading.Tasks.Task`1.Result"/> property returns a list of <see cref="VideoDevice">video devices</see>
+        /// that are currently available in the system.</returns>
+        public async Task<List<VideoDevice>> GetVideoDevicesAsync()
```
Comment (FabienLavocat) [inline src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs:189]: I like to return `IEnumerable<VideoDevice>` (personal preference)

File: src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs [hunk 11]
```diff
@@ -134,5 +179,45 @@ public async Task SetPreferredAudioOutputDeviceAsync(AudioDevice device)
...
+                }
+
+                return devices;
+            }).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Gets the video device that is currently used by the system.
+        /// </summary>
+        /// The <xref href="System.Threading.Tasks.Task`1.Result"/> property returns the <see cref="VideoDevice">video device</see>
```
Comment (FabienLavocat) [inline src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs:211]: Missing opening `<returns>`

File: src/DolbyIO.Comms.Sdk/Services/Video/LocalVideoService.cs [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+namespace DolbyIO.Comms
+{
+    /// <summary>
+    /// The local video service is used to control local participant's video capture
+    /// and sending into conference.
+    /// </summary>
+    public sealed class LocalVideoService
+    {
+        /// <summary>
+        /// /**
```
Comment (FabienLavocat) [inline src/DolbyIO.Comms.Sdk/Services/Video/LocalVideoService.cs:14]: Please remove this line.

**Review by FabienLavocat — CHANGES_REQUESTED** [review]

File: src/DolbyIO.Comms.Native/utils.h [hunk 13]
```diff
@@ -39,7 +39,7 @@ namespace dolbyio::comms::native {
   private:
     int result_;
   };
-
+
```
[the remark this answers is not in the document]
Comment (4thlabs) [reply src/DolbyIO.Comms.Native/utils.h:42]: Maybe I havn't seen something, why would I want to remove it ?

[the remark this answers is not in the document]
Comment (FabienLavocat) [reply src/DolbyIO.Comms.Native/utils.h:42]: Looks like you've added an extra space

File: src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs (hunk 4 above)
Comment (4thlabs) [reply src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs:24]: The code will look like thise `_sdk.MediaDevice.AudioAdded = () => {}`. When you read it, it is pretty straightforward. Adding Device will look like: `_sdk.MediaDevice.AudioDeviceAdded = () => {}` adding a repeating Device word. What do you think ?

Comment (FabienLavocat) [reply src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs:24]: Yes, I like it better (just my opinion), we can ask others what they think. This is consistent with our Web SDK: https://docs.dolby.io/communications-apis/docs/js-client-sdk-mediadeviceservice

File: src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs [hunk 14]
```diff
@@ -134,5 +179,45 @@ public async Task SetPreferredAudioOutputDeviceAsync(AudioDevice device)
...
+                }
+
+                return devices;
+            }).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Gets the video device that is currently used by the system.
+        /// </summary>
+        /// <returns> The <xref href="System.Threading.Tasks.Task`1.Result"/> property returns the <see cref="VideoDevice">video device</see>
```
Comment (FabienLavocat) [inline src/DolbyIO.Comms.Sdk/Services/MediaDeviceService.cs:211]: There is an extra space after `<returns>`

**Review by FabienLavocat — CHANGES_REQUESTED** [review]

**Review by srynk — APPROVED** [review]

File: src/DolbyIO.Comms.Native/utils.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (FabienLavocat) [reply src/DolbyIO.Comms.Native/utils.h:42]: @4thlabs

File: src/DolbyIO.Comms.Sdk/Native/Structs/VideoDevice.cs (hunk 3 above)
Comment (FabienLavocat) [reply src/DolbyIO.Comms.Sdk/Native/Structs/VideoDevice.cs:14]: @4thlabs

**Pull request merged**